src/metadata_directory/main.mo

```diff
@@ -14,83 +15,70 @@ actor class () = self {
+  let blobMap = OrderedMap.Make<Blob>(Blob.compare);
 
-  stable var stable_storage_map = storage_map.share();
-  stable var stable_wasm_modules_map = wasm_modules_map.share();
+  stable var storage : OrderedMap.Map<Principal, PrincipalRecord> = principalMap.empty<PrincipalRecord>();
 
   //
   // API for Metadata directory management
   //
 
   public shared query ({ caller }) func get_wasm_metadata() : async [WasmMetadata] {
-    let ?index = storage_map.get(caller) else return [];
-    Vec.get(storage, index) |> Vec.toArray(_);
+    let ?pr = principalMap.get(storage, caller) else return [];
+    pr.wasm_metadata_storage
+    |> blobMap.vals(_)
+    |> Iter.toArray(_);
   };
 
   public shared ({ caller }) func add_wasm_metadata(module_hash : Blob, description : ?Text, build_instructions : ?Text) : async () {
-    let wasm_modules = switch (storage_map.get(caller)) {
-      case (?index) Vec.get(storage, index);
+    let pr = switch (principalMap.get(storage, caller)) {
+      case (?value) value;
       case (null) {
-        let wasm_modules_new : Vec.Vector<WasmMetadata> = Vec.new();
-        storage_map.put(caller, Vec.size(storage));
-        Vec.add(storage, wasm_modules_new);
-        wasm_modules_new;
+        let pr_new : PrincipalRecord = {
+          var wasm_metadata_storage = blobMap.empty<WasmMetadata>();
+        };
+        storage := principalMap.put(storage, caller, pr_new);
+        pr_new;
       };
     };
     let is_valid = Blob.toArray(module_hash).size();
     if (is_valid != 32) throw Error.reject("Provided module hash is not valid.");
-    if (wasm_modules_map.get((caller, module_hash)) != null) throw Error.reject("The provided module hash already exists.");
-    wasm_modules_map.put((caller, module_hash), Vec.size(wasm_modules));
-    Vec.add(
-      wasm_modules,
+    if (blobMap.get(pr.wasm_metadata_storage, module_hash) != null) throw Error.reject("The provided module hash already exists.");
+    pr.wasm_metadata_storage := blobMap.put(
+      pr.wasm_metadata_storage,
+      module_hash,
       {
         module_hash;
         description = Option.get(description, "");
         build_instructions = Option.get(build_instructions, "");
         latest_update_timestamp = Prim.time();
+        created_timestamp = Prim.time();
       },
     );
   };
 
   public shared ({ caller }) func update_wasm_metadata(module_hash : Blob, description : ?Text, build_instructions : ?Text) : async () {
-    let ?directory_index = storage_map.get(caller) else throw Error.reject("There is no metadata for the wasm module.");
-    let ?index = wasm_modules_map.get((caller, module_hash)) else throw Error.reject("There is no metadata for the wasm module.");
+    let ?pr = principalMap.get(storage, caller) else throw Error.reject("There is no metadata for the wasm module.");
+    let ?wasm_metadata = blobMap.get(pr.wasm_metadata_storage, module_hash) else throw Error.reject("There is no metadata for the wasm module.");
```

Comment:
After this line the next check that comes, `is_valid != 32`, is superfluous. Because we only advance past this line if the module_hash is in the map, and it can only be in the map if it has the right length.

We can still do the length check but we should do above in the very first line of the function before we look anything up in any of the two maps.